### PR TITLE
Fixes #17376 - Compiler hangs when compiling inline recursive invocation #17376

### LINKED ISSUE #17376: Compiler hangs wne compiling inline recursive invocation
**Repro steps**
1. Create a new F# project with the following code in Program.fs:
``` F#
let rec inline test m = test m in test 0
```
2. Compile the project using `dotnet build`.

In fact, any recursive inline function can be declared and then invoked somewhere.

**Expected behavior**

Compiler is expected to output some error because inline rec functions are not allowed and quit.

**Actual behavior**

Compiler outputs several errors and warnings and then hangs with continuously increasing memory consumption. Compiler output before hang (with paths stripped):
```
Program.fs(1,25): error FS1114: The value 'test' was marked inline but was not bound in the optimization environment [TestApp.fsproj]
Program.fs(1,16): error FS1113: The value 'test' was marked inline but its implementation makes use of an internal or private function which is not sufficiently accessible [TestApp.fsproj]
Program.fs(1,25): warning FS1116: A value marked as 'inline' has an unexpected value [TestApp.fsproj]
Program.fs(1,25): warning FS1118: Failed to inline the value 'test' marked 'inline', perhaps because a recursive value was marked 'inline' [TestApp.fsproj]
Program.fs(1,25): error FS1114: The value 'test' was marked inline but was not bound in the optimization environment [TestApp.fsproj]
Program.fs(1,16): error FS1113: The value 'test' was marked inline but its implementation makes use of an internal or private function which is not sufficiently accessible [TestApp.fsproj]
Program.fs(1,25): error FS1114: The value 'test' was marked inline but was not bound in the optimization environment [TestApp.fsproj]
Program.fs(1,16): error FS1113: The value 'test' was marked inline but its implementation makes use of an internal or private function which is not sufficiently accessible [TestApp.fsproj]
```

N.B. This bug is contrived of course. I never thought of writing such code in real applications.

### PR DESCRIPTION
**Pull request by KevinRansom**
Fixes: https://github.com/dotnet/fsharp/issues/17376

Adds regression

### CODE REVIEW

File: docs/release-notes/.FSharp.Compiler.Service/9.0.100.md [hunk 1]
```diff
@@ -1,4 +1,5 @@
 ### Fixed
+* Compiler hangs wne compiling inline recursive invocation ([Issue #17376](https://github.com/dotnet/fsharp/issues/17376), [PR #17394](https://github.com/dotnet/fsharp/pull/17394))
```
Comment (slyshykO) [inline docs/release-notes/.FSharp.Compiler.Service/9.0.100.md:2]: typo **wne**
should be "when" ?

**Comment by T-Gro** [comment]
Could you also add the code snippet from the issue as a test?

**Comment by KevinRansom** [comment]
> Could you also add the code snippet from the issue as a test?



> Could you also add the code snippet from the issue as a test?

We have virtually identical code here:  https://github.com/dotnet/fsharp/blob/main/tests/FSharp.Compiler.ComponentTests/Conformance/InferenceProcedures/RecursiveSafetyAnalysis/E_RecursiveInline.fs

File: docs/release-notes/.FSharp.Compiler.Service/9.0.100.md (hunk 1 above)
Comment (KevinRansom) [reply docs/release-notes/.FSharp.Compiler.Service/9.0.100.md]: fixed

**Review by T-Gro — APPROVED** [review]

**Pull request merged**